File: packages/paso-py/paso/validator.py
import re
from urllib.parse import urlparse
from paso.types import PasoDeclaration, PasoCapability, ValidationError


VALID_METHODS = ['GET', 'POST', 'PUT', 'PATCH', 'DELETE']
VALID_PERMISSIONS = ['read', 'write', 'admin']
VALID_INPUT_TYPES = ['string', 'integer', 'number', 'boolean', 'enum', 'array', 'object']
VALID_OUTPUT_TYPES = ['string', 'integer', 'number', 'boolean', 'object', 'array']
VALID_AUTH_TYPES = ['api_key', 'bearer', 'oauth2', 'none']
VALID_IN_VALUES = ['query', 'path', 'body', 'header']
SNAKE_CASE_RE = re.compile(r'^[a-z][a-z0-9_]*$')


def validate(decl: PasoDeclaration) -> list[ValidationError]:
    """
    Validate a parsed PasoDeclaration against the spec.
    Returns a list of errors. Empty list = valid.
    """
    errors: list[ValidationError] = []

    # Version
    if not decl.version:
        errors.append(ValidationError(path='version', message='version is required'))
    elif decl.version != '1.0':
        errors.append(ValidationError(path='version', message='version must be "1.0"'))

    # Service
    if not decl.service:
        errors.append(ValidationError(path='service', message='service is required'))
    else:
        if not decl.service.name:
            errors.append(ValidationError(path='service.name', message='service.name is required'))
        if not decl.service.description:
            errors.append(ValidationError(path='service.description', message='service.description is required'))
        if not decl.service.base_url:
            errors.append(ValidationError(path='service.base_url', message='service.base_url is required'))
        else:
            if not _is_valid_url(decl.service.base_url):
                errors.append(ValidationError(path='service.base_url', message='service.base_url must be a valid URL'))

        if decl.service.auth:
            if decl.service.auth.type not in VALID_AUTH_TYPES:
                errors.append(ValidationError(
                    path='service.auth.type',
                    message=f'auth.type must be one of: {", ".join(VALID_AUTH_TYPES)}'
                ))

    # Capabilities
    if decl.capabilities is None:
        errors.append(ValidationError(path='capabilities', message='capabilities is required'))
    elif not isinstance(decl.capabilities, list):
        errors.append(ValidationError(path='capabilities', message='capabilities must be an array'))
    else:
        names: set[str] = set()
        for i, cap in enumerate(decl.capabilities):
            prefix = f'capabilities[{i}]'
            errors.extend(_validate_capability(cap, prefix, names))

    # Permissions
    if decl.permissions:
        cap_names = set(cap.name for cap in (decl.capabilities or []))
        all_referenced: set[str] = set()

        for tier in ['read', 'write', 'admin', 'forbidden']:
            tier_list = getattr(decl.permissions, tier, None)
            if tier_list:
                for name in tier_list:
                    # forbidden can reference capabilities not declared
                    if tier != 'forbidden' and name not in cap_names:
                        errors.append(ValidationError(
                            path=f'permissions.{tier}',
                            message=f'references unknown capability "{name}"'
                        ))
                    all_referenced.add(name)

        # Check forbidden doesn't overlap with tiers
        if decl.permissions.forbidden:
            tiered = set()
            if decl.permissions.read:
                tiered.update(decl.permissions.read)
            if decl.permissions.write:
                tiered.update(decl.permissions.write)
            if decl.permissions.admin:
                tiered.update(decl.permissions.admin)

            for name in decl.permissions.forbidden:
                if name in tiered:
                    errors.append(ValidationError(
                        path='permissions.forbidden',
                        message=f'"{name}" cannot be both in a permission tier and forbidden'
                    ))

    return errors


def _validate_capability(cap: PasoCapability, prefix: str, names: set[str]) -> list[ValidationError]:
    """
    Validate a single capability.
    """
    errors: list[ValidationError] = []

    if not cap.name:
        errors.append(ValidationError(path=f'{prefix}.name', message='name is required'))
    else:
        if not SNAKE_CASE_RE.match(cap.name):
            errors.append(ValidationError(
                path=f'{prefix}.name',
                message=f'"{cap.name}" must be snake_case'
            ))
        if cap.name in names:
            errors.append(ValidationError(
                path=f'{prefix}.name',
                message=f'duplicate capability name "{cap.name}"'
            ))
        names.add(cap.name)

    if not cap.description:
        errors.append(ValidationError(path=f'{prefix}.description', message='description is required'))

    if not cap.method:
        errors.append(ValidationError(path=f'{prefix}.method', message='method is required'))
    elif cap.method not in VALID_METHODS:
        errors.append(ValidationError(
            path=f'{prefix}.method',
            message=f'method must be one of: {", ".join(VALID_METHODS)}'
        ))

    if not cap.path:
        errors.append(ValidationError(path=f'{prefix}.path', message='path is required'))
    elif not cap.path.startswith('/'):
        errors.append(ValidationError(path=f'{prefix}.path', message='path must start with /'))

    if not cap.permission:
        errors.append(ValidationError(path=f'{prefix}.permission', message='permission is required'))
    elif cap.permission not in VALID_PERMISSIONS:
        errors.append(ValidationError(
            path=f'{prefix}.permission',
            message=f'permission must be one of: {", ".join(VALID_PERMISSIONS)}'
        ))

    # Validate inputs
    if cap.inputs:
        for input_name, input_obj in cap.inputs.items():
            input_prefix = f'{prefix}.inputs.{input_name}'
            if not input_obj.type:
                errors.append(ValidationError(path=input_prefix, message='type is required'))
            elif input_obj.type not in VALID_INPUT_TYPES:
                errors.append(ValidationError(
                    path=input_prefix,
                    message=f'type must be one of: {", ".join(VALID_INPUT_TYPES)}'
                ))

            if input_obj.type == 'enum' and (not input_obj.values or len(input_obj.values) == 0):
                errors.append(ValidationError(
                    path=input_prefix,
                    message='enum type must have values defined'
                ))

            if not input_obj.description:
                errors.append(ValidationError(path=input_prefix, message='description is required'))

            if input_obj.in_ and input_obj.in_ not in VALID_IN_VALUES:
                errors.append(ValidationError(
                    path=f'{input_prefix}.in',
                    message=f'in must be one of: {", ".join(VALID_IN_VALUES)}'
                ))

    # Validate path params exist in inputs
    if cap.path:
        path_params = re.findall(r'\{([^}]+)\}', cap.path)
        for param in path_params:
            if not cap.inputs or param not in cap.inputs:
                errors.append(ValidationError(
                    path=f'{prefix}.path',
                    message=f'path parameter "{{{param}}}" not found in inputs'
                ))
            elif cap.inputs and cap.inputs[param].in_ and cap.inputs[param].in_ != 'path':
                errors.append(ValidationError(
                    path=f'{prefix}.inputs.{param}',
                    message='path parameter must have in: path'
                ))

    # Validate output
    if cap.output:
        for field_name, output_obj in cap.output.items():
            if not output_obj.type:
                errors.append(ValidationError(
                    path=f'{prefix}.output.{field_name}',
                    message='type is required'
                ))
            elif output_obj.type not in VALID_OUTPUT_TYPES:
                errors.append(ValidationError(
                    path=f'{prefix}.output.{field_name}',
                    message=f'type must be one of: {", ".join(VALID_OUTPUT_TYPES)}'
                ))

    return errors


def _is_valid_url(url: str) -> bool:
    """
    Check if a string is a valid URL.
    """
    try:
        result = urlparse(url)
        return all([result.scheme, result.netloc])
    except Exception:
        return False

File: packages/paso-js/src/validator.ts
import { PasoDeclaration, PasoCapability, ValidationError } from './types';

const VALID_METHODS = ['GET', 'POST', 'PUT', 'PATCH', 'DELETE'];
const VALID_PERMISSIONS = ['read', 'write', 'admin'];
const VALID_INPUT_TYPES = ['string', 'integer', 'number', 'boolean', 'enum', 'array', 'object'];
const VALID_OUTPUT_TYPES = ['string', 'integer', 'number', 'boolean', 'object', 'array'];
const VALID_AUTH_TYPES = ['api_key', 'bearer', 'oauth2', 'none'];
const VALID_IN_VALUES = ['query', 'path', 'body', 'header'];
const SNAKE_CASE_RE = /^[a-z][a-z0-9_]*$/;
const MAX_NAME_LENGTH = 100;
const MAX_DESCRIPTION_LENGTH = 500;
const MAX_URL_LENGTH = 2000;

/**
 * Validate a parsed PasoDeclaration against the spec.
 * Returns an array of errors. Empty array = valid.
 */
export function validate(decl: PasoDeclaration): ValidationError[] {
  const errors: ValidationError[] = [];

  // Version
  if (!decl.version) {
    errors.push({ path: 'version', message: 'version is required' });
  } else if (decl.version !== '1.0') {
    errors.push({ path: 'version', message: 'version must be "1.0"' });
  }

  // Service
  if (!decl.service) {
    errors.push({ path: 'service', message: 'service is required' });
  } else {
    if (!decl.service.name) {
      errors.push({ path: 'service.name', message: 'service.name is required' });
    } else if (decl.service.name.length > MAX_NAME_LENGTH) {
      errors.push({
        path: 'service.name',
        message: `service.name exceeds ${MAX_NAME_LENGTH} characters`,
      });
    }
    if (!decl.service.description) {
      errors.push({ path: 'service.description', message: 'service.description is required' });
    } else if (decl.service.description.length > MAX_DESCRIPTION_LENGTH) {
      errors.push({
        path: 'service.description',
        message: `service.description exceeds ${MAX_DESCRIPTION_LENGTH} characters`,
      });
    }
    if (!decl.service.base_url) {
      errors.push({ path: 'service.base_url', message: 'service.base_url is required' });
    } else if (decl.service.base_url.length > MAX_URL_LENGTH) {
      errors.push({
        path: 'service.base_url',
        message: `service.base_url exceeds ${MAX_URL_LENGTH} characters`,
      });
    } else {
      try {
        const parsed = new URL(decl.service.base_url);
        if (parsed.protocol === 'http:') {
          errors.push({
            path: 'service.base_url',
            message: 'base_url uses http://. Consider https:// to protect auth tokens in transit',
            level: 'warning',
          });
        }
        if (isPrivateHost(parsed.hostname)) {
          errors.push({
            path: 'service.base_url',
            message:
              'base_url points to a private/internal address. If this YAML was shared with you, verify the URL before running usepaso serve.',
            level: 'warning',
          });
        }
      } catch {
        errors.push({ path: 'service.base_url', message: 'service.base_url must be a valid URL' });
      }
    }
    if (decl.service.auth) {
      if (!VALID_AUTH_TYPES.includes(decl.service.auth.type)) {
        errors.push({
          path: 'service.auth.type',
          message: `auth.type must be one of: ${VALID_AUTH_TYPES.join(', ')}`,
        });
      }
    }
  }

  // Capabilities
  if (!decl.capabilities) {
    errors.push({ path: 'capabilities', message: 'capabilities is required' });
  } else if (!Array.isArray(decl.capabilities)) {
    errors.push({ path: 'capabilities', message: 'capabilities must be an array' });
  } else {
    if (decl.capabilities.length === 0) {
      errors.push({
        path: 'capabilities',
        message: 'capabilities array is empty. MCP server will have no tools',
        level: 'warning',
      });
    }
    const names = new Set<string>();
    decl.capabilities.forEach((cap, i) => {
      const prefix = `capabilities[${i}]`;
      errors.push(...validateCapability(cap, prefix, names));
    });
  }

  // Permissions
  if (decl.permissions) {
    const capNames = new Set((decl.capabilities || []).map((c) => c.name));
    const allReferenced = new Set<string>();

    for (const tier of ['read', 'write', 'admin', 'forbidden'] as const) {
      const list = decl.permissions[tier];
      if (list && list.length === 0) {
        errors.push({
          path: `permissions.${tier}`,
          message: `empty array. Remove it or add capability names`,
          level: 'warning',
        });
      }
      if (list && list.length > 0) {
        for (const name of list) {
          // forbidden can reference capabilities not declared (to explicitly block API endpoints)
          if (tier !== 'forbidden' && !capNames.has(name)) {
            errors.push({
              path: `permissions.${tier}`,
              message: `references unknown capability "${name}"`,
            });
          }
          if (tier !== 'forbidden' && allReferenced.has(name)) {
            errors.push({
              path: `permissions.${tier}`,
              message: `"${name}" appears in multiple permission tiers`,
            });
          }
          allReferenced.add(name);
        }
      }
    }

    // Check forbidden doesn't overlap with tiers
    if (decl.permissions.forbidden) {
      const tiered = new Set([
        ...(decl.permissions.read || []),
        ...(decl.permissions.write || []),
        ...(decl.permissions.admin || []),
      ]);
      for (const name of decl.permissions.forbidden) {
        if (tiered.has(name)) {
          errors.push({
            path: 'permissions.forbidden',
            message: `"${name}" cannot be both in a permission tier and forbidden`,
          });
        }
      }
    }
  }

  return errors;
}

function validateCapability(
  cap: PasoCapability,
  prefix: string,
  names: Set<string>,
): ValidationError[] {
  const errors: ValidationError[] = [];

  if (!cap.name) {
    errors.push({ path: `${prefix}.name`, message: 'name is required' });
  } else {
    if (cap.name.length > MAX_NAME_LENGTH) {
      errors.push({
        path: `${prefix}.name`,
        message: `name exceeds ${MAX_NAME_LENGTH} characters`,
      });
    }
    if (!SNAKE_CASE_RE.test(cap.name)) {
      errors.push({ path: `${prefix}.name`, message: `"${cap.name}" must be snake_case` });
    }
    if (names.has(cap.name)) {
      errors.push({ path: `${prefix}.name`, message: `duplicate capability name "${cap.name}"` });
    }
    names.add(cap.name);
  }

  if (!cap.description) {
    errors.push({ path: `${prefix}.description`, message: 'description is required' });
  } else if (cap.description.length > MAX_DESCRIPTION_LENGTH) {
    errors.push({
      path: `${prefix}.description`,
      message: `description exceeds ${MAX_DESCRIPTION_LENGTH} characters`,
    });
  }

  if (!cap.method) {
    errors.push({ path: `${prefix}.method`, message: 'method is required' });
  } else if (!VALID_METHODS.includes(cap.method)) {
    errors.push({
      path: `${prefix}.method`,
      message: `method must be one of: ${VALID_METHODS.join(', ')}`,
    });
  }

  if (!cap.path) {
    errors.push({ path: `${prefix}.path`, message: 'path is required' });
  } else if (cap.path.length > MAX_URL_LENGTH) {
    errors.push({ path: `${prefix}.path`, message: `path exceeds ${MAX_URL_LENGTH} characters` });
  } else if (!cap.path.startsWith('/')) {
    errors.push({ path: `${prefix}.path`, message: 'path must start with /' });
  }

  if (!cap.permission) {
    errors.push({ path: `${prefix}.permission`, message: 'permission is required' });
  } else if (!VALID_PERMISSIONS.includes(cap.permission)) {
    errors.push({
      path: `${prefix}.permission`,
      message: `permission must be one of: ${VALID_PERMISSIONS.join(', ')}`,
    });
  }

  // Validate inputs
  if (cap.inputs) {
    for (const [inputName, input] of Object.entries(cap.inputs)) {
      const inputPrefix = `${prefix}.inputs.${inputName}`;
      if (!input.type) {
        errors.push({ path: inputPrefix, message: 'type is required' });
      } else if (!VALID_INPUT_TYPES.includes(input.type)) {
        errors.push({
          path: inputPrefix,
          message: `type must be one of: ${VALID_INPUT_TYPES.join(', ')}`,
        });
      }
      if (input.type === 'enum' && (!input.values || input.values.length === 0)) {
        errors.push({ path: inputPrefix, message: 'enum type must have values defined' });
      }
      if (!input.description) {
        errors.push({ path: inputPrefix, message: 'description is required' });
      } else if (input.description.length > MAX_DESCRIPTION_LENGTH) {
        errors.push({
          path: inputPrefix,
          message: `description exceeds ${MAX_DESCRIPTION_LENGTH} characters`,
        });
      }
      if (input.in && !VALID_IN_VALUES.includes(input.in)) {
        errors.push({
          path: `${inputPrefix}.in`,
          message: `in must be one of: ${VALID_IN_VALUES.join(', ')}`,
        });
      }
      if (input.in === 'body' && ['GET', 'DELETE'].includes(cap.method)) {
        errors.push({
          path: inputPrefix,
          message: `body parameters are not supported for ${cap.method} requests`,
        });
      }
    }
  }

  // Validate path params exist in inputs
  if (cap.path && cap.inputs) {
    const pathParams = cap.path.match(/\{([^}]+)\}/g);
    if (pathParams) {
      for (const param of pathParams) {
        const paramName = param.slice(1, -1);
        if (!cap.inputs[paramName]) {
          errors.push({
            path: `${prefix}.path`,
            message: `path parameter "{${paramName}}" not found in inputs`,
          });
        } else if (cap.inputs[paramName].in !== 'path') {
          errors.push({
            path: `${prefix}.inputs.${paramName}`,
            message: `path parameter must have in: path`,
          });
        }
      }
    }
  }

  // Validate constraints
  if (cap.constraints) {
    cap.constraints.forEach((c, ci) => {
      if (Object.keys(c).length === 0) {
        errors.push({
          path: `${prefix}.constraints[${ci}]`,
          message: 'empty constraint object. Add at least one field',
          level: 'warning',
        });
      }
    });
  }

  // Validate output
  if (cap.output) {
    for (const [fieldName, output] of Object.entries(cap.output)) {
      if (!output.type) {
        errors.push({ path: `${prefix}.output.${fieldName}`, message: 'type is required' });
      } else if (!VALID_OUTPUT_TYPES.includes(output.type)) {
        errors.push({
          path: `${prefix}.output.${fieldName}`,
          message: `type must be one of: ${VALID_OUTPUT_TYPES.join(', ')}`,
        });
      }
    }
  }

  return errors;
}

function isPrivateHost(hostname: string): boolean {
  const lower = hostname.toLowerCase().replace(/^\[|\]$/g, ''); // strip IPv6 brackets
  if (lower === 'localhost' || lower === '0.0.0.0') return true;

  // IPv6 private ranges
  if (lower === '::1' || lower === '0:0:0:0:0:0:0:1') return true; // loopback
  if (lower.startsWith('fe80:')) return true; // link-local
  if (lower.startsWith('fc00:') || lower.startsWith('fd00:')) return true; // unique local

  // IPv4 private ranges (reject leading zeros to block octal bypass like 0177.0.0.1)
  const rawParts = lower.split('.');
  const parts = rawParts.map((p) => (/^(0|[1-9]\d*)$/.test(p) ? parseInt(p, 10) : NaN));
  if (parts.length === 4 && parts.every((n) => !isNaN(n) && n >= 0 && n <= 255)) {
    const [a, b] = parts;
    if (a === 127) return true; // 127.x.x.x loopback
    if (a === 10) return true; // 10.x.x.x private
    if (a === 172 && b >= 16 && b <= 31) return true; // 172.16-31.x.x private
    if (a === 192 && b === 168) return true; // 192.168.x.x private
    if (a === 169 && b === 254) return true; // 169.254.x.x link-local / AWS metadata
  }

  return false;
}
